ipn/ipnlocal/network-lock.go

```diff
@@ -175,23 +175,24 @@ func (r *rotationTracker) addRotationDetails(np key.NodePublic, d *tka.RotationD
 // obsoleteKeys returns the set of node keys that are obsolete due to key rotation.
 func (r *rotationTracker) obsoleteKeys() set.Set[key.NodePublic] {
 	for _, v := range r.byWrappingKey {
+		// Do not consider signatures for keys that have been marked as obsolete
+		// by another signature.
+		v = slices.DeleteFunc(v, func(rd sigRotationDetails) bool {
+			return r.obsolete.Contains(rd.np)
+		})
+		if len(v) == 0 {
+			continue
+		}
+
 		// If there are multiple rotation signatures with the same wrapping
 		// pubkey, we need to decide which one is the "latest", and keep it.
 		// The signature with the largest number of previous keys is likely to
-		// be the latest, unless it has been marked as obsolete (rotated out) by
-		// another signature (which might happen in the future if we start
-		// compacting long rotated signature chains).
+		// be the latest.
 		slices.SortStableFunc(v, func(a, b sigRotationDetails) int {
-			// Group all obsolete keys after non-obsolete keys.
-			if ao, bo := r.obsolete.Contains(a.np), r.obsolete.Contains(b.np); ao != bo {
-				if ao {
-					return 1
-				}
-				return -1
-			}
 			// Sort by decreasing number of previous keys.
 			return b.numPrevKeys - a.numPrevKeys
 		})
+
 		// If there are several signatures with the same number of previous
 		// keys, we cannot determine which one is the latest, so all of them are
 		// rejected for safety.
```

Comment:
With truncation, if we have 2 chains of length 15, wouldn't this check mark them both as obsolete?
Or are we just relying on the fact that control will only return the single latest truncated chain?

Reply:
When we see a signature chain, we automatically mark all previous node keys found in the nested signatures (the "rotated away" keys) as obsolete, and ignore signatures for those node keys.

Control _does_ only return a single signature per node ID, but even if a node ends up seeing two signatures at the same time - before and after the rotation - the older signature will be marked as obsolete because the newer signature referred to it in its chain.

The test I added in this PR checks that after 100 rotations (when we've definitely reached the max chain length) the newest signature wins:
https://github.com/tailscale/tailscale/blob/56548ea538fc9c16563e3093e5d21565572df31a/ipn/ipnlocal/network-lock_test.go#L671-L694

Also an existing test checks that even a shorter signature chain wins over a longer one if it explicitly marks it as obsolete:
https://github.com/tailscale/tailscale/blob/56548ea538fc9c16563e3093e5d21565572df31a/ipn/ipnlocal/network-lock_test.go#L1278-L1287

(this is also why I propose we keep the chains, to answer your other question)